### PR TITLE
Finish Issue

### PR DESCRIPTION
**Pull request by ShaDy03**
I was tested all method (Add, get, getall, delete, update). Without error.

### CODE REVIEW

File: IssueTracker/DataAccess/Data/IData/IIssueData.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+namespace DataAccess.Data.IData
+{
+    public interface IIssueData
+    {
+        Task AddAsync(Issue entity);
+        Task<IEnumerable<Issue>> GetAllAsync();
+        Task<Issue?> GetByIdAsync(int id);
+        Task UpdateAsync(Issue entity);
+        Task DeleteAsync(int id);
+        Task NextPreview(int id, string method = "next");
```
Comment (iiusan) [inline IssueTracker/DataAccess/Data/IData/IIssueData.cs]: this method needs a better name

File: IssueTracker/DataAccess/Data/IData/IParticipantData.cs [hunk 2]
```diff
@@ -9,6 +9,6 @@ public interface IParticipantData
         Task<IEnumerable<Participant>> GetAllAsync();
         Task<Participant?> GetByIdAsync(int id);
         Task UpdateAsync(Participant participant);
-        Task<IEnumerable<Participant>> GetByProjectIdAsync(string storedProcedure, int id, string connectionId = "Default");
+        Task<IEnumerable<Participant>> GetByProjectIdAsync(int id, string storedProcedure = "dbo.spParticipant_GetAllByProjectId", string connectionId = "Default");
```
Comment (iiusan) [inline IssueTracker/DataAccess/Data/IData/IParticipantData.cs]: DB details, like stored procedures (or connections strings) should not be exposed here

File: IssueTracker/DataAccess/Models/Issue.cs [hunk 3]
```diff
@@ -22,11 +16,12 @@ public class Issue
         public int StatusId { get; set; }
         public int RoleId { get; set; }
         public bool IsDeleted { get; set; }
-        public IssueType? IssueType { get; set; }
-        public Priority? Priority { get; set; }
-        public Status? Status { get; set; }
-        public Role? Role { get; set; }
-        public Project? Project { get; set; }
+        public IssueType? IssueType { get; set; } = new IssueType();
```
Comment (iiusan) [inline IssueTracker/DataAccess/Models/Issue.cs]: doesn't look right. why is this marked optional and then instantiated?

File: IssueTracker/DataAccess/Data/IssueData.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                Updated = entity.Updated,
+                RoleId = entity.RoleId,
+                StatusId = entity.StatusId,
+                ProjectId = entity.ProjectId,
+                Description = entity.Description,
+                UserAssignedId = entity.UserAssignedId,
+                PriorityId = entity.PriorityId
+            });
+        }
+        public async Task<IEnumerable<Issue>> GetAllAsync() =>
```
Comment (iiusan) [inline IssueTracker/DataAccess/Data/IssueData.cs]: don't use expression body definitions

File: IssueTracker/IssueTrackerAPI/Controllers/IssueController.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+namespace IssueTrackerAPI.Controllers
+{
+    [Route("api")]
+    [ApiController]
+    public class IssueController : ControllerBase
+    {
+        private readonly IIssueData _issueDb;
+        private readonly IParticipantData _participantDb;
+        private readonly Mapper mapper;
+        public IssueController(IIssueData issueDb, IParticipantData participantData)
```
Comment (iiusan) [inline IssueTracker/IssueTrackerAPI/Controllers/IssueController.cs]: naming consistency

File: IssueTracker/IssueTrackerAPI/Controllers/IssueController.cs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            {
+                await _issueDb.NextPreview(id);
+                return Ok("Status update successful!");
+            }
+            return BadRequest("Error validation!");
+        }
+        
+        [Authorize]
+        [HttpPut("preview-Status")]
+        public async Task<IActionResult> PreviewStatusIssue(int id)
```
Comment (iiusan) [inline IssueTracker/IssueTrackerAPI/Controllers/IssueController.cs]: needs a better name. it's not clear what is going on here

File: IssueTracker/Validation/IssueValidation.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
+﻿using Models.Request;
+
+namespace Validation
+{
+    public class IssueValidation
+    {
+        public static bool IsValid(IssueRequest issue)
+        {
+            if(issue == null) return false;
+            if(issue.Title == String.Empty || issue.Title!.Length > 50) return false;
```
Comment (iiusan) [inline IssueTracker/Validation/IssueValidation.cs]: what happens if the Title is null?

File: IssueTracker/Validation/IssueValidation.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+            if (issue.Description == String.Empty) return false;
```
Comment (iiusan) [inline IssueTracker/Validation/IssueValidation.cs]: what happens if the Description is null?

File: IssueTracker/Validation/ProjectValidation.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
+﻿using Models.Request;
+
+namespace Validation
+{
+    public class ProjectValidation
+    {
+        public static bool IsValid(ProjectRequest project)
+        {
+            if(project == null) return false;
+            if(project.Title == String.Empty || project.Title!.Length > 50) return false;
```
Comment (iiusan) [inline IssueTracker/Validation/ProjectValidation.cs:10]: what happens if the title or description is null?

File: IssueTracker/Models/Request/IssueRequest.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
+﻿namespace Models.Request
+{
+    public class IssueRequest
```
Comment (iiusan) [inline IssueTracker/Models/Request/IssueRequest.cs:3]: initialize required properties using a constructor
https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/nullable-reference-types

File: IssueTracker/Models/Response/IssueResponse.cs [hunk 11]
```diff
@@ -2,5 +2,23 @@
 {
     public class IssueResponse
     {
+        public int Id { get; set; }
+        public int ProjectId { get; set; }
+        public string? Title { get; set; }
```
Comment (iiusan) [inline IssueTracker/Models/Response/IssueResponse.cs]: is the Title optional?

File: IssueTracker/IssueDb/dbo/StoredProcedures/Issue/spIssue_NextPreviewStatus.sql [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+begin
+	IF (@Method = 'next')
+	begin
+		if(@StatusId = 4)
+		begin
+			print 'This issue is max level of status!'
+		end
+		else
+		begin
+			Update Issue set StatusId = (@StatusId + 1) where Issue.Id = @Id
```
Comment (iiusan) [inline IssueTracker/IssueDb/dbo/StoredProcedures/Issue/spIssue_NextPreviewStatus.sql]: what happens if the status ids are out of order?

File: IssueTracker/DataAccess/DbAccess/SQLDataAccess.cs [hunk 13]
```diff
@@ -3,9 +3,13 @@
 using Dapper;
 using System.Data;
 using DataAccess.Utils;
+using System.Reflection;
 
 namespace DataAccess.DbAccess
 {
+#pragma warning disable CS8600 // Converting null literal or possible null value to non-nullable type.
```
Comment (iiusan) [inline IssueTracker/DataAccess/DbAccess/SQLDataAccess.cs]: remove these lines, and fix the warnings

File: IssueTracker/DataAccess/DbAccess/SQLDataAccess.cs [hunk 14]
```diff
@@ -14,24 +18,501 @@ public SQLDataAccess(IConfiguration config)
...
             string connectionId = "Default")
         {
             using (IDbConnection connection = new SqlConnection(_config.GetConnectionString(connectionId)))
             {
-                return await connection.QueryAsync<T>(storedProcedure, parameters,
+                return await connection.QueryAsync<TFirst>(storedProcedure, parameter,
+                    commandType: CommandType.StoredProcedure);
+            }
+        } //TODO: de gandit putin la aceste metode!
+        public async Task<IEnumerable<TFirst>> LoadDataAsync<TFirst, TSecond>(string storedProcedure, string connectionId = "Default")
```
Comment (iiusan) [inline IssueTracker/DataAccess/DbAccess/SQLDataAccess.cs:61]: can't everything be resolved with just one method? i.e. use a list of types and resolve with reflection?

File: IssueTracker/DataAccess/DataAccess.csproj [hunk 15]
```diff
@@ -9,6 +9,7 @@
   <ItemGroup>
     <PackageReference Include="Dapper" Version="2.0.123" />
     <PackageReference Include="Microsoft.Extensions.Configuration.Abstractions" Version="6.0.0" />
+    <PackageReference Include="Slapper.AutoMapper" Version="2.0.5" />
```
Comment (iiusan) [inline IssueTracker/DataAccess/DataAccess.csproj]: where is this used?

**Review by iiusan — CHANGES_REQUESTED** [review]

File: IssueTracker/DataAccess/Data/IData/IIssueData.cs (hunk 1 above)
Comment (ShaDy03) [reply IssueTracker/DataAccess/Data/IData/IIssueData.cs]: Change to NextStatusOfIssueAsync and PreviewStatusOfIssueAsync

File: IssueTracker/DataAccess/Data/IData/IParticipantData.cs (hunk 2 above)
Comment (ShaDy03) [reply IssueTracker/DataAccess/Data/IData/IParticipantData.cs]: Task<IEnumerable<Participant>> GetByProjectIdAsync(int id)

File: IssueTracker/DataAccess/Models/Issue.cs (hunk 3 above)
Comment (ShaDy03) [reply IssueTracker/DataAccess/Models/Issue.cs]: public IssueType IssueType { get; set; } = new IssueType();
        public Priority Priority { get; set; } = new Priority();
        public Status Status { get; set; } = new Status();
        public Role Role { get; set; } = new Role();
        public Project Project { get; set; } = new Project();
        public User User { get; set; } = new User();

File: IssueTracker/DataAccess/Data/IssueData.cs (hunk 4 above)
Comment (ShaDy03) [reply IssueTracker/DataAccess/Data/IssueData.cs]: i was resolved

File: IssueTracker/IssueTrackerAPI/Controllers/IssueController.cs (hunk 5 above)
Comment (ShaDy03) [reply IssueTracker/IssueTrackerAPI/Controllers/IssueController.cs]: public IssueController(IIssueData issue, IParticipantData participant)
        {
            _participant = participant;
            _issue = issue;
            mapper = AutoMapperConfig.Config();
        }

File: IssueTracker/Models/Response/IssueResponse.cs (hunk 11 above)
Comment (ShaDy03) [reply IssueTracker/Models/Response/IssueResponse.cs]: no, should i create a constructor with title parameter?

File: IssueTracker/IssueDb/dbo/StoredProcedures/Issue/spIssue_NextPreviewStatus.sql (hunk 12 above)
Comment (ShaDy03) [reply IssueTracker/IssueDb/dbo/StoredProcedures/Issue/spIssue_NextPreviewStatus.sql]: i change this procedure

File: IssueTracker/DataAccess/DataAccess.csproj (hunk 15 above)
Comment (ShaDy03) [reply IssueTracker/DataAccess/DataAccess.csproj]: i remove this nuget

File: IssueTracker/DataAccess/DbAccess/SQLDataAccess.cs (hunk 13 above)
Comment (ShaDy03) [reply IssueTracker/DataAccess/DbAccess/SQLDataAccess.cs]: I will try to solve this.

File: IssueTracker/DataAccess/DbAccess/SQLDataAccess.cs (hunk 14 above)
Comment (ShaDy03) [reply IssueTracker/DataAccess/DbAccess/SQLDataAccess.cs:61]: I will try to solve this.

File: IssueTracker/DataAccess/Data/IData/IIssueData.cs (hunk 1 above)
Comment (iiusan) [reply IssueTracker/DataAccess/Data/IData/IIssueData.cs]: rename Preview to Previous

File: IssueTracker/DataAccess/Models/Issue.cs (hunk 3 above)
Comment (iiusan) [reply IssueTracker/DataAccess/Models/Issue.cs]: why is instantiation needed here?

File: IssueTracker/DataAccess/Data/IData/IIssueData.cs (hunk 1 above)
Comment (ShaDy03) [reply IssueTracker/DataAccess/Data/IData/IIssueData.cs]: i change in PreviousStatusOfIssueAsync

File: IssueTracker/IssueTrackerAPI/Controllers/IssueController.cs (hunk 6 above)
Comment (ShaDy03) [reply IssueTracker/IssueTrackerAPI/Controllers/IssueController.cs]: [HttpPut("previousStatusIssue")]
        public async Task<IActionResult> PreviousStatusIssue(int id)

**Review by iiusan — APPROVED** [review]

**Pull request merged**